src/main/java/org/paulnikepro/hw4/todoapp/service/TodoServiceImpl.java

```diff
@@ -0,0 +1,158 @@
+package org.paulnikepro.hw4.todoapp.service;
+
+import lombok.RequiredArgsConstructor;
+import org.paulnikepro.hw4.todoapp.dto.TodoCreateDto;
+import org.paulnikepro.hw4.todoapp.dto.TodoResponseDto;
+import org.paulnikepro.hw4.todoapp.dto.TodoUpdateDto;
+import org.paulnikepro.hw4.todoapp.model.TaskHistory;
+import org.paulnikepro.hw4.todoapp.model.Todo;
+import org.paulnikepro.hw4.todoapp.mapper.TaskHistoryMapper;
+import org.paulnikepro.hw4.todoapp.mapper.TodoMapper;
+import org.paulnikepro.hw4.todoapp.model.enums.Status;
+import org.paulnikepro.hw4.todoapp.repository.TaskHistoryRepository;
+import org.paulnikepro.hw4.todoapp.repository.TodoRepository;
+import jakarta.transaction.Transactional;
+import org.paulnikepro.hw4.todoapp.dto.TaskHistoryResponseDto;
+import org.paulnikepro.hw4.todoapp.exception.ResourceNotFoundException;
+import org.springframework.stereotype.Service;
+
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+@RequiredArgsConstructor
+@Service
+public class TodoServiceImpl implements TodoService {
+
+    private final TodoRepository todoRepository;
+    private final TodoMapper todoMapper;
+    private final TaskHistoryRepository taskHistoryRepository;
+    private final TaskHistoryMapper taskHistoryMapper;
+
+    @Override
+    public void deleteById(Long id) {
+        todoRepository.deleteById(id);
+    }
+
+    @Override
+    public boolean existsById(Long id) {
+        return todoRepository.existsById(id);
+    }
+
+    public TodoResponseDto findById(Long id) {
+        Optional<Todo> todoOptional = todoRepository.findById(id);
+
+        return todoOptional.map(this::convertToDto).orElse(null);
```

Comment:
It is bad practice to use Optional and return null. Better to throw ResourceNotFoundException in this case